### PR TITLE
`getResultSet` should return null after `executeUpdate`

### LINKED ISSUE #3523: Create table sql not works
## Bug Report
```java
Exception in thread "main" java.lang.NullPointerException
	at org.apache.shardingsphere.core.execute.sql.execute.result.StreamQueryResult.<init>(StreamQueryResult.java:49)
	at org.apache.shardingsphere.shardingjdbc.jdbc.core.statement.ShardingPreparedStatement.getResultSet(ShardingPreparedStatement.java:136)
	at org.jetbrains.exposed.sql.Transaction$exec$2.executeInternal(Transaction.kt:107)
	at org.jetbrains.exposed.sql.statements.Statement.executeIn$exposed(Statement.kt:59)
	at org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:183)
	at org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:125)
	at org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:101)
	at org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:92)
	at com.xiachufang.DbUtilsKt.dropTables(DbUtils.kt:22)
	at com.xiachufang.DatasourceKt$main$1.invoke(Datasource.kt:51)
	at com.xiachufang.DatasourceKt$main$1.invoke(Datasource.kt)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt$inTopLevelTransaction$1.invoke(ThreadLocalTransactionManager.kt:156)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt$inTopLevelTransaction$2.invoke(ThreadLocalTransactionManager.kt:197)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.keepAndRestoreTransactionRefAfterRun(ThreadLocalTransactionManager.kt:205)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.inTopLevelTransaction(ThreadLocalTransactionManager.kt:196)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt$transaction$1.invoke(ThreadLocalTransactionManager.kt:134)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.keepAndRestoreTransactionRefAfterRun(ThreadLocalTransactionManager.kt:205)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.transaction(ThreadLocalTransactionManager.kt:106)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.transaction(ThreadLocalTransactionManager.kt:104)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.transaction$default(ThreadLocalTransactionManager.kt:103)
	at com.xiachufang.DatasourceKt.main(Datasource.kt:50)
	at com.xiachufang.DatasourceKt.main(Datasource.kt)
```

### Which version of ShardingSphere did you use?
4.0.0-RC2 4.0.0-RC3

### Which project did you use? Sharding-JDBC or Sharding-Proxy?
Sharding-JDBC

### Expected behavior
```sql
CREATE TABLE IF NOT EXISTS board (id BIGINT PRIMARY KEY, author_id BIGINT NOT NULL, title VARCHAR(100) NOT NULL, description VARCHAR(1000) NOT NULL, status INT NOT NULL, accessibility INT NOT NULL, create_time DATETIME(6) NOT NULL, update_time DATETIME(6) NOT NULL)
```

Failed.

### Actual behavior
It should create the table.

### Reason analyze (If you can)
`StreamQueryResult`  did not consider `resultSet` to be null. When executing create table sql, resultSet will be null.

```java
public final class StreamQueryResult implements QueryResult {
    
    private final QueryResultMetaData metaData;
    
    private final ResultSet resultSet;
    
    @SneakyThrows
    public StreamQueryResult(final ResultSet resultSet, final ShardingRule shardingRule) {
        this.resultSet = resultSet;
        metaData = new QueryResultMetaData(resultSet.getMetaData(), shardingRule);
    }
    
    @SneakyThrows
    public StreamQueryResult(final ResultSet resultSet, final EncryptRule encryptRule) {
        this.resultSet = resultSet;
        metaData = new QueryResultMetaData(resultSet.getMetaData(), encryptRule);
    }
    
    @SneakyThrows
    public StreamQueryResult(final ResultSet resultSet) {
        this.resultSet = resultSet;
        metaData = new QueryResultMetaData(resultSet.getMetaData());
    }
    
```

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
Execute the following sql should reproduce the problem.

```sql
CREATE TABLE IF NOT EXISTS board (id BIGINT PRIMARY KEY, author_id BIGINT NOT NULL, title VARCHAR(100) NOT NULL, description VARCHAR(1000) NOT NULL, status INT NOT NULL, accessibility INT NOT NULL, create_time DATETIME(6) NOT NULL, update_time DATETIME(6) NOT NULL)
```

### PR DESCRIPTION
**Pull request by gfreezy**
Fixes #3523.

According to the doc of  interface `Statement`, `Statement.getResultSet` should return `null` when `the result is an update count`.

```java

    /**
     *  Retrieves the current result as a <code>ResultSet</code> object.
     *  This method should be called only once per result.
     *
     * @return the current result as a <code>ResultSet</code> object or
     * <code>null</code> if the result is an update count or there are no more results
     * @exception SQLException if a database access error occurs or
     * this method is called on a closed <code>Statement</code>
     * @see #execute
     */
    ResultSet getResultSet() throws SQLException;

```

### CODE REVIEW

File: sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingStatement.java [hunk 1]
```diff
@@ -113,7 +113,19 @@ public ResultSet getResultSet() throws SQLException {
         for (Statement each : statementExecutor.getStatements()) {
             ResultSet resultSet = each.getResultSet();
             resultSets.add(resultSet);
-            queryResults.add(new StreamQueryResult(resultSet, connection.getRuntimeContext().getRule(), connection.getRuntimeContext().getProps(), sqlRouteResult.getSqlStatementContext()));
+            if (resultSet != null) {
+                queryResults.add(new StreamQueryResult(resultSet, connection.getRuntimeContext().getRule(), connection.getRuntimeContext().getProps(), sqlRouteResult.getSqlStatementContext()));
+            }
+        }
+        boolean allResultSetAreNull = true;
```
Comment (SteNicholas) [inline sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingStatement.java]: 这个判断可以优化下吧？

File: sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java [hunk 2]
```diff
@@ -133,7 +133,19 @@ public ResultSet getResultSet() throws SQLException {
         for (Statement each : preparedStatementExecutor.getStatements()) {
             ResultSet resultSet = each.getResultSet();
             resultSets.add(resultSet);
-            queryResults.add(new StreamQueryResult(resultSet, connection.getRuntimeContext().getRule(), connection.getRuntimeContext().getProps(), sqlRouteResult.getSqlStatementContext()));
+            if (resultSet != null) {
+                queryResults.add(new StreamQueryResult(resultSet, connection.getRuntimeContext().getRule(), connection.getRuntimeContext().getProps(), sqlRouteResult.getSqlStatementContext()));
+            }
+        }
+        boolean allResultSetAreNull = true;
```
Comment (SteNicholas) [inline sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java]: 这个判断可以优化下吧？

Comment (gfreezy) [reply sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java]: 可以用 stream 吗？sharding-jdbc 最低支持 java 什么版本？

Comment (SteNicholas) [reply sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java]: 你这个放在上面的循环里不就行了。

Comment (gfreezy) [reply sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java]: 已修复。

Comment (gfreezy) [reply sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java]: 上一个 commit 的 ci 为啥挂了。没看出啥原因

**Comment by terrymanu** [comment]
@SteNicholas I approve it, anything else for you?

File: sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java [hunk 3]
```diff
@@ -130,10 +130,17 @@ public ResultSet getResultSet() throws SQLException {
...
         for (Statement each : preparedStatementExecutor.getStatements()) {
             ResultSet resultSet = each.getResultSet();
             resultSets.add(resultSet);
-            queryResults.add(new StreamQueryResult(resultSet, connection.getRuntimeContext().getRule(), connection.getRuntimeContext().getProps(), sqlRouteResult.getSqlStatementContext()));
+            if (resultSet != null) {
+                queryResults.add(new StreamQueryResult(resultSet, connection.getRuntimeContext().getRule(), connection.getRuntimeContext().getProps(), sqlRouteResult.getSqlStatementContext()));
+                hasResultSets = true;
+            }
+        }
+        if (!hasResultSets) {
```
Comment (SteNicholas) [inline sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java]: Directly check by whether queryResults is empty?

File: sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingStatement.java [hunk 4]
```diff
@@ -110,10 +110,17 @@ public ResultSet getResultSet() throws SQLException {
...
         for (Statement each : statementExecutor.getStatements()) {
             ResultSet resultSet = each.getResultSet();
             resultSets.add(resultSet);
-            queryResults.add(new StreamQueryResult(resultSet, connection.getRuntimeContext().getRule(), connection.getRuntimeContext().getProps(), sqlRouteResult.getSqlStatementContext()));
+            if (resultSet != null) {
+                queryResults.add(new StreamQueryResult(resultSet, connection.getRuntimeContext().getRule(), connection.getRuntimeContext().getProps(), sqlRouteResult.getSqlStatementContext()));
+                hasResultSets = true;
+            }
+        }
+        if (!hasResultSets) {
```
Comment (SteNicholas) [inline sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingStatement.java]: Directly check by whether queryResults is empty?

**Comment by SteNicholas** [comment]
> @SteNicholas I approve it, anything else for you?

I have some comments which need author optimize.

**Comment by gfreezy** [comment]
> > @SteNicholas I approve it, anything else for you?
> 
> I have some comments which need author optimize.

resolved

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatementTest.java [hunk 5]
```diff
@@ -31,6 +31,7 @@
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertNull;
```
Comment (SteNicholas) [inline sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatementTest.java]: This import position should be above the import of aasertThat.

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingStatementTest.java [hunk 6]
```diff
@@ -29,6 +29,7 @@
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertNull;
```
Comment (SteNicholas) [inline sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingStatementTest.java]: 同上

**Comment by SteNicholas** [comment]
@gfreezy Please update import order.

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatementTest.java (hunk 5 above)
Comment (gfreezy) [reply sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatementTest.java]: Is there any documentation stating import orders ? I didn't find any.

**Comment by gfreezy** [comment]
@SteNicholas Done. But if there is an import order to ensure, we should add an import order lint.

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatementTest.java (hunk 5 above)
Comment (SteNicholas) [reply sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatementTest.java]: No, but this import orders are always according to character order.

**Comment by SteNicholas** [comment]
@terrymanu LGTM, please merge.

**Comment by SteNicholas** [comment]
@gfreezy Check this CI failure result.

**Comment by gfreezy** [comment]
@SteNicholas It seems the problem of CI. How can i rerun it.

**Comment by SteNicholas** [comment]
@gfreezy Please close this pull request, and reopen this pr.

**Comment by nancyzrh** [comment]
please use '/run ci' to restart ci, it is tested ok.

**Comment by terrymanu** [comment]
Hi, the exception is caused by NPE, maybe we should debug the exception, not just run ci without any changes

File: sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java [hunk 7]
```diff
@@ -133,7 +133,12 @@ public ResultSet getResultSet() throws SQLException {
         for (Statement each : preparedStatementExecutor.getStatements()) {
             ResultSet resultSet = each.getResultSet();
             resultSets.add(resultSet);
-            queryResults.add(new StreamQueryResult(resultSet, connection.getRuntimeContext().getRule(), connection.getRuntimeContext().getProps(), sqlRouteResult.getSqlStatementContext()));
+            if (resultSet != null) {
+                queryResults.add(new StreamQueryResult(resultSet, connection.getRuntimeContext().getRule(), connection.getRuntimeContext().getProps(), sqlRouteResult.getSqlStatementContext()));
+            }
+        }
+        if (queryResults.isEmpty()) {
+            return null;
```
Comment (SteNicholas) [inline sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java]: 这边直接返回null的话会影响GeneralDQLIT的assertExecuteForStatement报空指针异常，你这边应该返回queryResults

**Comment by SteNicholas** [comment]
@gfreezy Please modify the return value according above comment.

File: sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java (hunk 7 above)
Comment (gfreezy) [reply sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java]: 这里必须要返回 `ResultSet` 类型，但是我没办法构造一个空的 `ResultSet`。

java 文档里面写的是应该要返回 null。

https://docs.oracle.com/javase/7/docs/api/java/sql/Statement.html#getResultSet()

Comment (SteNicholas) [reply sharding-jdbc/sharding-jdbc-core/src/main/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingPreparedStatement.java]: 那就不要在这边直接返回return null呗，也就是把这个判断return null去掉。

**Comment by terrymanu** [comment]
@gfreezy @SteNicholas hi, Please use English to discuss code review suggestions.

**Comment by gfreezy** [comment]
@terrymanu @SteNicholas https://github.com/apache/incubator-shardingsphere/pull/3547/commits/ce3be4be7e04f921f2b92ec590510b616a27640e  
This change can bypass the null check. But i think it's really not a good idea to write like that. It is hard to get what will be returned.

**Pull request merged**